### PR TITLE
Code cleanup for ITMApplication

### PR DESCRIPTION
**Pull request by tcobbs-bentley**
Also contains updates to other classes that are a direct result of `ITMApplication` changes.

Notes:

1. The `ITMMessenger` changes were done to avoid potential issues that previously required the suppression of `LeakingThis`. Changing `logger` to `observable` and updating the messenger when the logger changes maintains the old functionality.
2. The `ITMNativeUI` changes were also done to avoid having to suppress `LeakingThis`, and to avoid running into problems due to that suppression.
3. The `ITMWebViewLogger` change was due to fixes added to `ITMApplication` to detach the logger when appropriate.

### CODE REVIEW

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt [hunk 1]
```diff
@@ -185,25 +191,31 @@ open class ITMApplication(
...
+     * code). The default logger uses [Log][android.util.Log] for the messages. Replace this object
+     * with an [ITMLogger] subclass to change the logging behavior.
+     *
+     * > __Note:__ Updating this value also updates the logger being used by [messenger].
      */
-    @Suppress("LeakingThis")
-    var messenger = ITMMessenger(this)
+    var logger: ITMLogger by Delegates.observable(ITMLogger()) { _, _, new ->
+        messenger.logger = new
+    }
```
Comment (toddsouthenbentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt:202]: Does observable provide any advantage over using a property setter? 

```suggestion
    var logger = ITMLogger()
        set (value) {
            field = value
            messenger.logger = value
        }
```

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMNativeUI.kt [hunk 2]
```diff
@@ -61,12 +61,16 @@ open class ITMNativeUI(
...
+    val components: MutableList<ITMNativeUIComponent> = mutableListOf()
 
-    init {
-        @Suppress("LeakingThis")
+    /**
+     * Register the standard [ITMNativeUIComponent] subclasses that are part of mobile-sdk-android.
+     *
+     * This is called automatically by [ITMApplication.createNativeUI].
+     */
+    fun registerStandardComponents() {
```
Comment (toddsouthenbentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMNativeUI.kt:72]: Can you please explain how were you able to remove the @Suppress("LeakingThis") lines? I don't see any changes to  `ITMNativeUI` that seem to facilitate it, so I'm missing something more subtle.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt [hunk 3]
```diff
@@ -558,22 +602,23 @@ open class ITMApplication(
...
+                } else {
+                    super.shouldOverrideUrlLoading(view, request)
                 }
-                return super.shouldOverrideUrlLoading(view, request)
             }
 
             override fun onRenderProcessGone(view: WebView, detail: RenderProcessGoneDetail): Boolean {
-                if (this@ITMApplication.onRenderProcessGone(view, detail)) {
-                    return true
+                return if (this@ITMApplication.onRenderProcessGone(view, detail)) {
```
Comment (toddsouthenbentley) [inline mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt:617]: Again, this should work:

```
return this@ITMApplication.onRenderProcessGone(view, detail) ||
       super.onRenderProcessGone(view, detail)
```

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt (hunk 1 above)
Comment (tcobbs-bentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt:202]: Not that I know of, but the same question could be asked in reverse, and the observable is a bit more succinct.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt [hunk 4]
```diff
@@ -558,22 +602,23 @@ open class ITMApplication(
...
             override fun shouldInterceptRequest(view: WebView, request: WebResourceRequest): WebResourceResponse? {
-                val result = this@ITMApplication.shouldInterceptRequest(view, request)
-                return result ?: super.shouldInterceptRequest(view, request)
+                return this@ITMApplication.shouldInterceptRequest(view, request) ?: super.shouldInterceptRequest(view, request)
             }
 
             override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
-                if (this@ITMApplication.shouldOverrideUrlLoading(view, request)) {
-                    return true
+                return if (this@ITMApplication.shouldOverrideUrlLoading(view, request)) {
```
[the remark this answers is not in the document]
Comment (tcobbs-bentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt]: Fixed.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt (hunk 3 above)
Comment (tcobbs-bentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt]: Fixed.

File: mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt (hunk 1 above)
Comment (tcobbs-bentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt:202]: I don't mind changing it to a set if you really feel that is better, but I personally don't see either one as being better or worse.

Comment (toddsouthenbentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt:202]: I don't have a strong opinion either way. Using a setter is more common I think. I'd never heard of observable before this.

Comment (tcobbs-bentley) [reply mobile-sdk/src/main/java/com/github/itwin/mobilesdk/ITMApplication.kt:202]: I only learned about observable when I was investigating lazy, but felt that it was appropriate here. Given no strong feeling either way, I'm going to leave this as it is.

**Review by toddsouthenbentley — APPROVED** [review]

**Pull request merged**